### PR TITLE
gregord integration, high-level flow

### PR DESCRIPTION
**Pull request by patrickxb**
r? @maxtaco

### CODE REVIEW

File: gregord/config.go [hunk 1]
```diff
@@ -241,6 +246,7 @@ func parseOptions(argv []string, quiet bool) (*Options, error) {
 	fs.StringVar(&raw.bindAddress, "bind-address", os.Getenv("BIND_ADDRESS"), "hostname:port to bind to")
 	fs.StringVar(&raw.mysqlDSN, "mysql-dsn", os.Getenv("MYSQL_DSN"), "user:pw@host/dbname for MySQL")
 	fs.BoolVar(&raw.debug, "debug", false, "turn on debugging")
+	fs.BoolVar(&raw.mockAuth, "mock-auth", false, "turn on mock authentication")
```
Comment (maxtaco) [inline gregord/config.go]: this would be a very bad thing to enable by accident

File: rpc/events.go [hunk 2]
```diff
@@ -4,10 +4,50 @@ import (
 	"github.com/keybase/gregor/protocol/gregor1"
 )
 
-type eventHandler interface {
-	connectionCreated(gregor1.UID)
-	connectionDestroyed(gregor1.UID)
-	uidServerCreated(gregor1.UID)
-	uidServerDestroyed(gregor1.UID)
-	broadcastSent(gregor1.Message)
+type EventHandler interface {
```
Comment (maxtaco) [inline rpc/events.go]: can this file go into the test/ directory?

File: storage/sql_sm.go [hunk 3]
```diff
@@ -28,6 +28,14 @@ func NewSQLEngine(d *sql.DB, of gregor.ObjFactory, stw sqlTimeWriter, cl clockwo
 	return &SQLEngine{driver: d, objFactory: of, stw: stw, clock: cl}
 }
 
+func NewMySQLEngine(d *sql.DB, of gregor.ObjFactory) *SQLEngine {
+	return NewSQLEngine(d, of, mysqlTimeWriter{}, clockwork.NewRealClock())
+}
+
+func NewSqlLiteSQLEngin(d *sql.DB, of gregor.ObjFactory) *SQLEngine {
```
Comment (maxtaco) [inline storage/sql_sm.go]: check spelling...

**Comment by maxtaco** [comment]
Change `.travis.yml` to reflect the new environment variable for enabling MYSQL.

**Comment by maxtaco** [comment]
(Also, see the C/I bugs above..)

File: gregord/consumer.go [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+		return err
+	}
+	c.db = db
+	return nil
+}
+
+func (c *consumer) forceParseTime(dsn *url.URL) *url.URL {
+	// We need to have parseTime=true as one of our DSN paramenters
+	query := dsn.Query()
+	query.Set("parseTime", "true")
```
Comment (maxtaco) [inline gregord/consumer.go]: There's code in `sql_test` that does this, can we merge this code with that code?

File: gregord/config.go (hunk 1 above)
Comment (patrickxb) [inline gregord/config.go]: It should be removed in the near-future, but I needed this to make `gregord` work.

File: rpc/events.go (hunk 2 above)
Comment (patrickxb) [inline rpc/events.go]: the interface needs to be with `rpc` package, but I can try moving implementation to `test`.  Will also have to move `flow_test.go` there.

Comment (maxtaco) [inline rpc/events.go]: There's other code that's shared and test only that I keep in `test/`.

File: rpc/server.go [hunk 5]
```diff
@@ -232,7 +236,11 @@ func (s *Server) consume(c context.Context, m gregor1.Message) error {
 	retCh := make(chan error)
 	args := messageArgs{c, m, retCh}
 	s.consumeCh <- args
-	return <-retCh
+	if err := <-retCh; err != nil {
+		return err
+	}
+	s.broadcastCh <- args
```
Comment (maxtaco) [inline rpc/server.go]: we'll eventually have a third step once we have pubsub

File: .travis.yml [hunk 6]
```diff
@@ -14,7 +14,7 @@ addons:
 before_install:
   - mysql -e "create database IF NOT EXISTS test;" -uroot
 env:
-  - MYSQL_DSN=root:@/test GO15VENDOREXPERIMENT=1
+  - TEST_MYSQL_DSN=root:@/test GO15VENDOREXPERIMENT=1
```
Comment (maxtaco) [inline .travis.yml]: do we still need the vendoring directive now that we have 1.6?

Comment (patrickxb) [inline .travis.yml]: No, I think @jacobhaven mentioned that a while ago...

Comment (maxtaco) [inline .travis.yml]: I thought so too, I'm surprised to still see it here

File: gregord/flow_test.go [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+
+	// wait for the broadcast
+	<-events.BcastSent
+
+	// check that all the clients received the message
+	for i := 0; i < len(clients); i++ {
+		if len(clients[i].broadcasts) != 1 {
+			t.Errorf("clients[%d] broadcasts: %d, expected 1", i, len(clients[i].broadcasts))
+		}
+	}
```
Comment (maxtaco) [inline gregord/flow_test.go]: can you also test that clients you start up after the fact can catch up?

File: .travis.yml (hunk 6 above)
Comment (jacobhaven) [inline .travis.yml]: Yep it's unnecessary; delete it.

Comment (patrickxb) [inline .travis.yml]: Just pushed a commit with it deleted.

File: gregord/consumer.go (hunk 4 above)
Comment (patrickxb) [inline gregord/consumer.go]: done.

File: gregord/main.go [hunk 8]
```diff
@@ -19,15 +21,46 @@ func main() {
...
 
-	log.Fatal(newMainServer(opts, srv).listenAndServe())
+func setupMockAuth(srv *grpc.Server) {
+	var auth mockAuth
+	srv.SetAuthenticator(auth)
+}
+
+type mockAuth int
+
+func (m mockAuth) AuthenticateSessionToken(_ context.Context, tok gregor1.SessionToken) (gregor1.AuthResult, error) {
```
Comment (jacobhaven) [inline gregord/main.go]: `tok` should also be `_`

File: gregord/main.go [hunk 9]
```diff
@@ -19,15 +21,46 @@ func main() {
...
+func setupSessionAuth(opts *Options, srv *grpc.Server) *grpc.SessionCacher {
+	conn, err := opts.SessionServer.Dial()
+	if err != nil {
+		log.Fatal(err)
+	}
 	Cli := rpc.NewClient(rpc.NewTransport(conn, nil, keybase1.WrapError), keybase1.ErrorUnwrapper{})
 	sc := grpc.NewSessionCacher(gregor1.AuthClient{Cli}, clockwork.NewRealClock(), 10*time.Minute)
-	defer sc.Close()
-	srv.SetAuthenticator(sc)
+	srv.SetAuthenticator(grpc.NewSessionCacher(gregor1.AuthClient{Cli}, clockwork.NewRealClock(), 10*time.Minute))
```
Comment (jacobhaven) [inline gregord/main.go]: `srv.SetAuthenticator(sc)`

File: gregord/main.go [hunk 10]
```diff
@@ -19,15 +21,46 @@ func main() {
 	}
 
 	srv := grpc.NewServer()
-	conn, err := opts.SessionServer.Dial()
+
+	if opts.MockAuth {
+		setupMockAuth(srv)
```
Comment (jacobhaven) [inline gregord/main.go]: `srv.SetAuthenticator(mockAuth{})`

File: gregord/main.go (hunk 9 above)
Comment (patrickxb) [inline gregord/main.go]: thanks, I had a messy rebase.

File: gregord/flow_test.go (hunk 7 above)
Comment (maxtaco) [inline gregord/flow_test.go]: (this would test that things are being persisted to the DB properly)

**Comment by maxtaco** [comment]
Just saw another c/I failure: 

{code}
2016/04/13 21:33:00 opening mysql connection to root:@/test?parseTime=true
--- FAIL: TestConsumeBroadcastFlow (0.00s)
	flow_test.go:104: dial tcp 127.0.0.1:19911: getsockopt: connection refused
{code}

**Comment by patrickxb** [comment]
Yes, I saw that already...will fix.

File: gregord/flow_test.go (hunk 7 above)
Comment (patrickxb) [inline gregord/flow_test.go]: @jacobhaven implementing client rpc for InbandMessagesSince.  I will add a ticket for adding such a test after both PRs are merged.

**Comment by patrickxb** [comment]
@maxtaco PTAL

CI green, the other remaining test dependent on a different PR.

**Comment by maxtaco** [comment]
Did you try some stress-testing and/or testing with artificial sleeps in key locations?  Would be nice to ferret out any flakey tests before the PR goes in.

**Comment by patrickxb** [comment]
I ran stress.bash on it with and without race detection.

**Comment by maxtaco** [comment]
Ok great, can you also try some deadlockers, especially around the TCP port listen / dial stuff?

**Comment by patrickxb** [comment]
Done

**Comment by maxtaco** [comment]
ok :+1:

**Pull request merged**